### PR TITLE
Add hyperdual numbers

### PR DESCRIPTION
**Pull request by Aertalon**
WIP, needs a bit more testing and also extending math functions (needs second order derivative)

### CODE REVIEW

File: src/math.hpp [hunk 1]
```diff
@@ -5,79 +5,65 @@
...
+#include <iostream>
 #include <type_traits>
 
 namespace opt {
 
-template <Arithmetic T>  // clang-format off
-  requires(not Dual<T>)
-constexpr auto exp(T x)  // clang-format on
+template <class T>  // clang-format off
+  requires(Arithmetic<T> || Dual<T>)
```
Comment (oliverlee) [inline src/math.hpp:14]: template `<Arithmetic T>`
is sufficient a `Dual` is also an `Arithmetic`

File: src/math.hpp [hunk 2]
```diff
@@ -5,79 +5,65 @@
...
     auto acc = decltype(x){1};
     auto power = acc;
     auto factorial = acc;
 
-    constexpr std::size_t taylor_terms{10U};
+    constexpr std::size_t taylor_terms{15U};
     for (std::size_t i{1U}; i < taylor_terms; ++i) {
-        power *= x;
-        factorial *= static_cast<decltype(factorial)>(i);
+        power = power * x;
```
Comment (oliverlee) [inline src/math.hpp:23]: Are these compound operators not implemented yet?

File: src/math.hpp [hunk 3]
```diff
@@ -5,79 +5,65 @@
...
+        factorial = factorial * T{static_cast<float>(i)};
```
Comment (oliverlee) [inline src/math.hpp:24]: Should we implement implicit construction of a `dual<Arithmetic>` from `int`? The conversion to `float` may not work for a user-defined type.

**Review by oliverlee — COMMENTED** [review]
nice

File: src/math.hpp (hunk 1 above)
Comment (Aertalon) [reply src/math.hpp:14]: Done

File: src/math.hpp (hunk 2 above)
Comment (Aertalon) [reply src/math.hpp:23]: Done

File: src/math.hpp (hunk 3 above)
Comment (Aertalon) [reply src/math.hpp:24]: Right. I can do that tomorrow

**Comment by oliverlee** [comment]
Should I rebase the geometric series stuff on this PR?

File: test/dualnumbers_test.cpp [hunk 4]
```diff
@@ -10,28 +10,44 @@ auto main() -> int
     using opt::dual;
 
     test("dualnumbers sum") = [] {
-        constexpr dual x{1.0F, 0.0F};
-        constexpr dual y{0.0F, 1.0F};
+        constexpr dual x{1.0F, 0.0F, 0.0F, 0.0F};
```
[the remark this answers is not in the document]
Comment (Aertalon) [reply test/dualnumbers_test.cpp:13]: I renamed the elements e1, e2, e3 in the latest patch - we can change them to i,j,k, they just look too much like quaternions to me, but it's a weak argument. I'm ok with the function objects. Would it be possible to name them e1, e2, e3 and write something like e1 + 2.F*e2 and have them automatically figure out that we want them to have floats as duals?

File: src/convopt.hpp [hunk 5]
```diff
@@ -33,14 +33,46 @@ constexpr auto gradient(const P& p, F cost) -> distance_t<P>
...
+        auto d = opt::point<dual<T>, N>{};
+
+        for (std::size_t i{0U}; i < N; ++i) {
+            d[i].real = p[i];
+        }
+
+        return d;
+    }();
+
+    auto h = std::array<distance_t<P>, N>{};
```
[the remark this answers is not in the document]
Comment (Aertalon) [reply src/convopt.hpp:60]: Yes, an Hessian should be captured in some sort of matrix, similarly to how a gradient is capture in a vector/distance element. This is just to have a rough idea of what this method returns

File: test/dualnumbers_test.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Aertalon) [reply test/dualnumbers_test.cpp:13]: On the other hand, the user should not ever see/use dual numbers, these appear here only to test that they work as expected. Not sure

**Comment by oliverlee** [comment]
Did you rebase off `origin/main`? It's weird that there is a rebase commit.

**Review by oliverlee — APPROVED** [review]

**Pull request merged**